core/modules/taxonomy/taxonomy.admin.inc

```diff
@@ -17,64 +17,74 @@ function taxonomy_overview_vocabularies($form) {
   $term_counts = db_query("SELECT vocabulary, count(tid) FROM {taxonomy_term_data} GROUP BY vocabulary")->fetchAllKeyed();
   foreach ($vocabularies as $vocabulary) {
     $terms_count = array_key_exists($vocabulary->machine_name, $term_counts) ? $term_counts[$vocabulary->machine_name] : 0;
-    $form[$vocabulary->machine_name]['#vocabulary'] = $vocabulary;
-    $form[$vocabulary->machine_name]['name'] = array('#markup' => check_plain($vocabulary->name));
-    $form[$vocabulary->machine_name]['items'] = array('#markup' => $terms_count);
-    $form[$vocabulary->machine_name]['weight'] = array(
-      '#type' => 'weight',
-      '#title' => t('Weight for @title', array('@title' => $vocabulary->name)),
-      '#title_display' => 'invisible',
-      '#delta' => 10,
-      '#default_value' => $vocabulary->weight,
-    );
-    $links = array();
-    $links['list'] = array(
-      'title' => t('List terms'),
-      'href' => "admin/structure/taxonomy/$vocabulary->machine_name",
-    );
-    $links['add'] = array(
-      'title' => t('Add terms'),
-      'href' => "admin/structure/taxonomy/$vocabulary->machine_name/add",
-    );
-    $links['configure'] = array(
-      'title' => t('Configure'),
-      'href' => "admin/structure/taxonomy/$vocabulary->machine_name/configure",
-    );
-    if (module_exists('field_ui') && user_access('administer fields')) {
-      $links['fields'] = array(
-        'title' => t('Manage fields'),
-        'href' => "admin/structure/taxonomy/$vocabulary->machine_name/fields",
-        'weight' => 5,
+    if (taxonomy_vocabulary_access(FALSE, $vocabulary)) {
+      $form[$vocabulary->machine_name]['#vocabulary'] = $vocabulary;
+      $form[$vocabulary->machine_name]['name'] = array('#markup' => check_plain($vocabulary->name));
+      $form[$vocabulary->machine_name]['items'] = array('#markup' => $terms_count);
+      if (user_access('administer taxonomy')) {
+        $form[$vocabulary->machine_name]['weight'] = array(
+          '#type' => 'weight',
+          '#title' => t('Weight for @title', array('@title' => $vocabulary->name)),
+          '#title_display' => 'invisible',
+          '#delta' => 10,
+          '#default_value' => $vocabulary->weight,
+        );
+      }
+      $links = array();
+      $links['list'] = array(
+        'title' => t('List terms'),
+        'href' => "admin/structure/taxonomy/$vocabulary->machine_name",
       );
-      $links['display'] = array(
-        'title' => t('Manage displays'),
-        'href' => "admin/structure/taxonomy/$vocabulary->machine_name/display",
-        'weight' => 10,
+      if (taxonomy_vocabulary_access('create', $vocabulary)) {
+        $links['add'] = array(
+          'title' => t('Add terms'),
+          'href' => "admin/structure/taxonomy/$vocabulary->machine_name/add",
+        );
+      }
+      if (user_access('administer taxonomy')) {
```

Comment:
It feels that this needs to be checking per-vocabulary permission instead.

Reply:
There is no per-vocab permission as far as I know.